aidial_adapter_vertexai/dial_api/request.py

```diff
@@ -39,3 +48,28 @@ def get_attachments(message: Message) -> List[Attachment]:
     if custom_content is None:
         return []
     return custom_content.attachments or []
+
+
+def collect_text_content(
```

Comment:
Inside of bedrock we don't have this `assert_never` advantage, but

```python
def collect_text_content(
    content: MessageContentSpecialized, delimiter: str = "\n\n"
) -> str:

    if content is None:
        return ""

    if isinstance(content, str):
        return content

    texts: List[str] = []
    for part in content:
        if isinstance(part, MessageContentTextPart):
            texts.append(part.text)
        else:
            raise ValidationError(
                "Can't extract text from a multi-modal content part"
            )

    return delimiter.join(texts)
```

this function has max 2 tab levels

this `match/case` approach has 4 levels, it becomes much harder to read this

Reply:
Conceptually it's much easier because it deals with each of the two type unions uniformly and exhaustively. 

Once you see this `match-case-assert_never` pattern, you don't need to keep track of all of the if-then-else-return cases which **supposed** to cover all of the possible cases. 

This pattern **does** cover all of the cases and the handlers are all presented on the same level, which makes it easier to read and write. Besides, it's statically verified.

This pattern is purely mechanistic, it provides a scaffolding for arbitrary data transformations. You only need to think about handlers in particular places in the data structure. The control flow on the big scale is defined by this recursive pattern matching, which eases the cognitive load. 

I wish I did it in bedrock :(